tests/test_pipeline.py

```diff
@@ -0,0 +1,30 @@
+from select import select
+
+import pytest
+
+from psycopg import waiting
+
+pytestmark = pytest.mark.libpq(">= 14")
+
+
+@pytest.mark.slow
+def test_pipeline_communicate(pgconn, demo_pipeline, generators):
```

Comment:
Agreed. These were necessary for me to come up with a working algorithm but I've now moved them as scripts.
Incidentally, running `python tests/scripts/pipeline-demo.py  -n 1 [--async]` produces a deadlock, so there seem to be a problem in the high level API (no problem with the generator AFAICT): to be investigated...

Reply:
> Incidentally, running `python tests/scripts/pipeline-demo.py -n 1 [--async]` produces a deadlock, so there seem to be a problem in the high level API

Strangely, the error is gone now after rebase. Perhaps because the waiting function (`wait_async()`) changed in the meantime.

Reply:
Hm, now it fails with `python tests/scripts/pipeline-demo.py  -n10 [--async]`; investigating...